lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1723,6 +1763,177 @@ class SCFToCalyxPass : public circt::impl::SCFToCalyxBase<SCFToCalyxPass> {
 private:
   LogicalResult partialPatternRes;
   std::shared_ptr<calyx::CalyxLoweringState> loweringState = nullptr;
+
+  /// Creates a new new top-level function based on `baseName`.
+  FuncOp createNewTopLevelFn(ModuleOp moduleOp, std::string &baseName) {
+    std::string newName = baseName;
+    unsigned counter = 0;
+    while (SymbolTable::lookupSymbolIn(moduleOp, newName)) {
+      newName = llvm::join_items("_", baseName, std::to_string(++counter));
+    }
+
+    OpBuilder builder(moduleOp.getContext());
+    builder.setInsertionPointToStart(moduleOp.getBody());
+
+    FunctionType funcType = builder.getFunctionType({}, {});
+
+    if (auto newFunc =
+            builder.create<FuncOp>(moduleOp.getLoc(), newName, funcType)) {
+      baseName = newName;
+      return newFunc;
+    }
+
+    moduleOp.emitError("Cannot create new top-level function.");
+
+    return nullptr;
+  }
+
+  /// Insert a call from the newly created top-level function/`caller` to the
+  /// old top-level function/`callee`; and create `memref.alloc`s inside the new
+  /// top-level function for arguments with `memref` types and for the
+  /// `memref.alloc`s inside `callee`.
+  void insertCallFromNewTopLevel(OpBuilder &builder, FuncOp caller,
+                                 FuncOp callee) {
+    if (caller.getBody().empty()) {
+      caller.addEntryBlock();
+    }
+
+    Block *callerEntryBlock = &caller.getBody().front();
+    builder.setInsertionPointToStart(callerEntryBlock);
+
+    // For those non-memref arguments passing to the original top-level
+    // function, we need to copy them to the new top-level function.
+    SmallVector<Type, 4> nonMemRefCalleeArgTypes;
+    for (auto arg : callee.getArguments()) {
+      if (!isa<MemRefType>(arg.getType())) {
+        nonMemRefCalleeArgTypes.push_back(arg.getType());
+      }
+    }
+
+    for (Type type : nonMemRefCalleeArgTypes) {
+      callerEntryBlock->addArgument(type, caller.getLoc());
+    }
+
+    FunctionType callerFnType = caller.getFunctionType();
+    SmallVector<Type, 4> updatedCallerArgTypes(
+        caller.getFunctionType().getInputs());
+    updatedCallerArgTypes.append(nonMemRefCalleeArgTypes.begin(),
+                                 nonMemRefCalleeArgTypes.end());
+    caller.setType(FunctionType::get(caller.getContext(), updatedCallerArgTypes,
+                                     callerFnType.getResults()));
+
+    Block *calleeFnBody = &callee.getBody().front();
+    unsigned originalCalleeArgNum = callee.getArguments().size();
+
+    SmallVector<Value, 4> extraMemRefArgs;
+    SmallVector<Type, 4> extraMemRefArgTypes;
+    SmallVector<Value, 4> extraMemRefOperands;
+    SmallVector<Operation *, 4> opsToModify;
+    for (auto &op : callee.getBody().getOps()) {
+      if (isa<memref::AllocaOp>(op) || isa<memref::AllocOp>(op) ||
+          isa<memref::GetGlobalOp>(op))
```

Comment:
```suggestion
      if (isa<memref::AllocaOp, memref::AllocOp, memref::GetGlobalOp>(op))
```

^ I think this should work.